### PR TITLE
Try many solvers

### PR DESCRIPTION
**Pull request by abakst**
This is a proposal for throwing multiple solvers at a problem in offline mode.

The changes are:

- Allow the user to specify a _list_ of solvers (or the string "all") to try for each goal
- For each goal, `forkIO` a thread for each solver. If a thread returns a definite result, then use that as the result for the goal. Otherwise, each thread has either timed out or thrown an exception, in which case we might want to report those exceptions, for example (this is what the PR does currently).

At this point I'm looking for comments/suggestions.

### CODE REVIEW

**Comment by robdockins** [comment]
You might take a look at how I did portfolio solving in cryptol:
https://github.com/GaloisInc/cryptol/blob/3bec6cdbb4e01701c467a76931dbb0cb00d558e9/src/Cryptol/Symbolic/What4.hs#L448

I think using the `async` library for this simplifies things a lot.

**Comment by abakst** [comment]
Yes that looks much simpler

File: crux/src/Crux/Goal.hs [hunk 1]
```diff
@@ -237,33 +239,66 @@ proveGoalsOffline adapter opts ctx explainFailure (Just gs0) = do
...
+          mapM_ kill as'
+          return $ Right r
+        Right r@(NotProved _ (Just _)) -> do
+          mapM_ kill as'
+          return $ Right r
+        Right _ ->
+          await as' es
+
+    -- `cancel` from async blocks until the canceled thread has terminated.
+    kill a = throwTo (asyncThreadId a) AsyncCancelled
```
Comment (robdockins) [inline crux/src/Crux/Goal.hs:301]: Don't use an async exception here.  This generally bypasses exception handling code; but we want exception-handling to send signals to shutdown external solvers and have them shutdown.  `AsyncCancelled` will leave the solvers running and eating CPU time.  I tend to use `ExitSuccess`.

File: crux/src/Crux/Goal.hs [hunk 2]
```diff
@@ -237,33 +239,66 @@ proveGoalsOffline adapter opts ctx explainFailure (Just gs0) = do
...
+                    Sat (evalFn, _) -> do
+                      let model = ctx ^. cruciblePersonality . personalityModel
+                      vals <- evalModel evalFn model
+                      explain <- explainFailure evalFn p
+                      return (NotProved explain (Just (ModelView vals)))
+                    Unknown -> do
+                      return (NotProved mempty Nothing)
+                case mres of
+                  Just res -> return res
+                  Nothing -> return (NotProved mempty Nothing)
```
Comment (robdockins) [inline crux/src/Crux/Goal.hs:260]: The nothing case corresponds to timeout, right? Probably worth mentioning that in the explanation.

File: crux/src/Crux/Goal.hs [hunk 3]
```diff
@@ -237,33 +239,66 @@ proveGoalsOffline adapter opts ctx explainFailure (Just gs0) = do
...
-                return (Prove (p, gt))
-              Unknown -> do
-                end
-                modifyIORef' goalNum (updateProcessedGoals p (NotProved mempty Nothing))
-                return (Prove (p, NotProved mempty Nothing))
-          case mres of
-            Just res -> return res
-            Nothing -> return (Prove (p, NotProved mempty Nothing))
 
+          let runOneSolver adapter = do
```
Comment (robdockins) [inline crux/src/Crux/Goal.hs:243]: I'm not a big fan of complicated nestings of `where` and `let`.  If we can do it without too much difficulty, I'd like to see this lifted to the same `where` block as `go` (which could really use a better name).

File: crux/src/Crux/Config/Solver.hs [hunk 4]
```diff
@@ -65,6 +66,9 @@ data SolverConfig = SingleOnlineSolver SolverOnline
                   | OnlyOfflineSolver SolverOffline
                   -- ^ Use only an offline solver with no support for path
                   -- satisfiability checking
+                  | OnlyOfflineSolvers [SolverOffline]
```
Comment (robdockins) [inline crux/src/Crux/Config/Solver.hs:69]: Does `OnlyOfflineSolver` need to be a separate case?

File: crux/src/Crux/Goal.hs (hunk 2 above)
Comment (abakst) [reply crux/src/Crux/Goal.hs:260]: Do you happen to know if the timeout corresponds to wall clock or actual CPU time? That is, is a per-thread timeout the same behavior as a a timeout on the `waitAnyCatch`? Not terribly important, but I think `waitAnyCatch` ends up catching the `Timeout` exception, which means that if `withTimeout` returns, it only returns a `Just` value.

File: crux/src/Crux/Config/Solver.hs (hunk 4 above)
Comment (abakst) [reply crux/src/Crux/Config/Solver.hs:69]: Not necessarily, I'm ambivalent about whether or not this should be a separate case or to just use pattern matching. I can remove the constructor.

File: crux/src/Crux/Goal.hs (hunk 2 above)
Comment (robdockins) [reply crux/src/Crux/Goal.hs:260]: I think it's wall-clock time, but I'm not sure.  Given the way this is written, I'd expect the `withTimeout` combinator to handle the `Timeout` exception before it bubbles out, but I don't know offhand where that combinator comes from or how it works, exactly.

**Comment by robdockins** [comment]
I'm worried this is starting to fall behind master.  Is there still work to do before merging this?

**Comment by abakst** [comment]
I'm not planning on making any more changes, so if it looks good I think we can merge it.

**Review by robdockins — APPROVED** [review]
I'm happy with this as long as CI gives the thumbs up

**Pull request merged**